### ISSUE TITLE
Error handling

### ISSUE THREAD

**Issue by prepor** [c-cube/cconv#4]
Hello. Thank you for a great job!

Do you have any ideas about better error handling? Now cconv just says something like  "conversion error: unexpected  string" for any conversation error of any depth in nested document.

I think there are two possible solutions:
1. Use exceptions to reconstruct path on error case
2. pass some _context_ object through all decoders
#2 can be more "functional" and explicit, but #1 doesn't touch signatures and, I think, can be more efficient.

**Comment by c-cube** [comment]
That's an interesting idea indeed. I also think 1) is the way to go, but I lack time to add it to cconv nowadays. Contributing would be very welcome!

edit: that would also be the occasion of making a new release…

**Comment by prepor** [comment]
I'm glad to contribute, but can't run tests :)

After `ocaml setup.ml -configure --enable-yojson --enable-ppx --enable-bencode --enable-sexp --enable-tests`

`make` command fails with 

```
+ /Users/prepor/.opam/condo/bin/ocamlfind ocamlc -c -g -annot -bin-annot -ppxopt ppx_deriving,ppx/ppx_deriving_cconv.cma -package yojson -package ppx_deriving.std -package oUnit -package bencode -I tests -o tests/run_tests_ppx.cmo tests/run_tests_ppx.ml
File "tests/run_tests_ppx.ml", line 5, characters 2-27:
Error: Unbound module CConv
```

There is also tags like `use_ppx_cconv, pkg_cconv, pkg_cconv.yojson, pkg_cconv.bencode` in _tags file, but where are they should be defined (is this oasis magic?). There are also warnings:

```
Warning: the tag "pkg_cconv.bencode" is not used in any flag or dependency declaration, so it will have no effect; it may be a typo. Otherwise you can use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
File "_tags", line 64, characters 46-62:
Warning: the tag "pkg_cconv.yojson" is not used in any flag or dependency declaration, so it will have no effect; it may be a typo. Otherwise you can use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
File "_tags", line 64, characters 35-44:
Warning: the tag "pkg_cconv" is not used in any flag or dependency declaration, so it will have no effect; it may be a typo. Otherwise you can use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
Configuration "true: quiet, debug, tests", line 1, characters 20-25:
Warning: the tag "tests" is not used in any flag or dependency declaration, so it will have no effect; it may be a typo. Otherwise you can use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
```

**Comment by c-cube** [comment]
If you want to test, I suggest you make sure it compiles, and use `opam pin -k git cconv .#HEAD` so that the installed version is your development version. It is simpler imho than trying to use a ppx outside the scope of ocamlfind.

Also we'd need to upgrade the ppx to the most recent version of `ppx_deriving`…

**Comment by prepor** [comment]
After pin I have this error:

```
+ /Users/prepor/.opam/4.03.0/bin/ocamlfind ocamldep -package cconv -package cconv.bencode -package cconv.yojson -package yojson -package ppx_deriving.std -package oUnit -package bencode -modules tests/run_tests_ppx.ml > tests/run_tests_ppx.ml.depends
File "tests/run_tests_ppx.ml", line 5, characters 21-26:
Cannot locate deriver cconv
```

Is this ppx_deriving version issue? But everything works fine in utop, for example.

I'm sorry, for these silly questions, i'm just trying to reconstruct build chain :(

**Comment by c-cube** [comment]
Don't apologize, I'm myself sorry that it's so complicated, and I didn't touch `ppx_deriving` plugins for a while :/

**Comment by prepor** [comment]
Ok, after I had replaced magic tags in _tags by "package(cconv.ppx), package(cconv.yojson), package(cconv.bencode)" tests compiled. But not passed ;)

```
ocaml setup.ml -test
......FFFFFFF........F
==============================================================================
Failure: cconv:6:record_ignore

expected: { "x": 1 } but got: { "x": 1, "y": 2 }
------------------------------------------------------------------------------
==============================================================================
Failure: cconv:3::0:@encoder

expected: { "i": { "bint": 1 }, "j": 42 } but got: { "i": 1, "j": 42 }
------------------------------------------------------------------------------
==============================================================================
Failure: cconv:2:record_encoder:5:bij_bencode

conversion error: unexpected  int
------------------------------------------------------------------------------
==============================================================================
Failure: cconv:2:record_encoder:4:bij_json

conversion error: unexpected  int
------------------------------------------------------------------------------
==============================================================================
Failure: cconv:2:record_encoder:3:bij_bencode

conversion error: unexpected  int
------------------------------------------------------------------------------
==============================================================================
Failure: cconv:2:record_encoder:2:bij_json

conversion error: unexpected  int
------------------------------------------------------------------------------
==============================================================================
Failure: cconv:2:record_encoder:1:bij_bencode

conversion error: unexpected  int
------------------------------------------------------------------------------
==============================================================================
Failure: cconv:2:record_encoder:0:bij_json

conversion error: unexpected  int
------------------------------------------------------------------------------
Ran: 22 tests in: 0.00 seconds.
FAILED: Cases: 22 Tried: 22 Errors: 0 Failures: 8 Skip:  0 Todo: 0 Timeouts: 0.
```

**Comment by c-cube** [comment]
Please commit the new `_tags` file, it's nice to have fixed the tests! So apparently the system of annotations (e.g. `[@ignore]`) is broken (i.e. the tags are ignored), which makes some tests fail. Or maybe it's because `ppx_deriving` changed a while ago; can you try to modify the test file to change, say, `y: int [@ignore]` into `y: (int [@ignore])`?

**Comment by prepor** [comment]
Yep, 

```
type boxed_int = {
  bint : (int [@encoder CConv.Encode.(map box_int encode_boxed_int)]) [@ignore];
}
```

In this source `@encoder` now is type attribute and `@ignore` is label attribute. 

How should it be fixed? Looks like it logical, that `@encoder`/`@decoder` attributes are type attributes, but `@ignore` (as `@default` and `@key` which I want to add) are label attributes, but it brokes API.

**Comment by c-cube** [comment]
Does it work with this parenthesing? The location of ppx attributes changed in 4.02.3, I think, and `ppx_deriving` had to break as well.

**Comment by prepor** [comment]
Yes, this parenthesing moves attributes from label level to type, as ppx_deriving_cconv expects.

**Comment by c-cube** [comment]
closed by #6.
